### PR TITLE
[12.0][ADD] - Add new addon: contract_forecast

### PR DESCRIPTION
**Pull request by sbejaoui**

### CODE REVIEW

**Comment by pedrobaeza** [comment]
I see more interesting to do this through `mis_builder_forecast` module that is in PR, creating forecast entries there. What do you think?

**Comment by sbidoul** [comment]
@pedrobaeza that's an interesting idea. You mean mis_builder_cash_flow? https://github.com/OCA/mis-builder/pull/154

Here we additional data that does not fit in the forecast table of `mis_builder_cash_flow` (such as start/end date, the link to the contract etc).

At first sight, I would not create such a dependency.

Perhaps we could create a glue module to populate the cash flow forecast table when the contract_forecast changes. That should take into account the payment terms, because here our goal is to forecast revenue (invoice dates), not cash.

File: contract_forecast/hooks.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    Generate contract line forecast periods
+    """
+    _logger.info(
+        "Post init hook for module post_init_hook: "
+        "Generate contract line forecast periods"
+    )
+    with api.Environment.manage():
+        env = api.Environment(cr, SUPERUSER_ID, {})
+        contract_lines = env["account.analytic.invoice.line"].search([])
+        contract_lines.with_delay()._generate_forecast_periods()
```
Comment (sbidoul) [inline contract_forecast/hooks.py]: @sbejaoui this creates one job that will take ages to run in a database loaded with many lines.

You could improve it in several ways:
- search where is_canceled = False
- search where auto_renew is False and end_date is null or in the future
- create one job per line (or per contract)

File: contract_forecast/models/contract_line.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        res = super(AccountAnalyticInvoiceLine, self).write(values)
+        if any(
+            [
+                field in values
+                for field in self._get_forecast_update_trigger_fields()
+            ]
+        ):
+            for rec in self:
+                rec._unlink_forecast_periods()
+                rec._generate_forecast_periods()
```
Comment (sbidoul) [inline contract_forecast/models/contract_line.py]: @sbejaoui the unlink must be inside generate_forecast_period, to be part of the same transaction.

Also, there is a with_delay missing here, no?

File: contract_forecast/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
+# Copyright 2019 ACSONE SA/NV
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+{
+    "name": "Contract Forecast",
+    "description": """
```
Comment (nikul-serpentcs) [inline contract_forecast/__manifest__.py]: @sbejaoui @aheficent `description` - > `summary`

File: contract_forecast/models/contract_line.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+class AccountAnalyticInvoiceLine(models.Model):
+
+    _inherit = "account.analytic.invoice.line"
+
+    forecast_period_ids = fields.One2many(
+        comodel_name="contract.line.forecast.period",
+        inverse_name="contract_line_id",
+        string="Forecast Periods",
+        required=False,
```
Comment (nikul-serpentcs) [inline contract_forecast/models/contract_line.py:19]: @sbejaoui @aheficent no need `required=False`, By default `False`

**Review by nikul-serpentcs — COMMENTED** [review]
Nitpicking

File: contract_forecast/models/contract_line.py (hunk 4 above)
Comment (ThomasBinsfeld) [reply contract_forecast/models/contract_line.py:19]: @nikul-serpentcs Code is never too explicit, let's keep it as it :slightly_smiling_face:

File: contract_forecast/__manifest__.py (hunk 3 above)
Comment (ThomasBinsfeld) [reply contract_forecast/__manifest__.py]: @nikul-serpentcs Done

**Comment by ThomasBinsfeld** [comment]
The module is now rebased on the 12.0 branch and compatible with it (split from analytic account).

**Comment by sbejaoui** [comment]
@ThomasBinsfeld , can you add queue to oca_dependencies.txt

**Review by ThomasBinsfeld — APPROVED** [review]

**Review by gva-acsone — APPROVED** [review]
Functional test +1

**Review by rousseldenis — APPROVED** [review]
LGTM

File: setup/contract_forecast/.eggs/README.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+This directory contains eggs that were downloaded by setuptools to build, test, and run plug-ins.
+
+This directory caches those eggs to prevent repeated downloads.
+
+However, it is safe to delete this directory.
+
```
Comment (sbidoul) [inline setup/contract_forecast/.eggs/README.txt]: this .eggs directory must be removed

File: contract_forecast/readme/DESCRIPTION.rst [hunk 6]
```diff
@@ -0,0 +1 @@
+This module add the possibility to analyse contract forecast.
```
Comment (sbidoul) [inline contract_forecast/readme/DESCRIPTION.rst:1]: A `USAGE.rst` would be nice.

**Review by sbidoul — COMMENTED** [review]
@OCA/project-service-maintainers if there are no objections we are going to merge this soon.

**Pull request merged**